### PR TITLE
The WebRTC PR to close them all

### PR DESCRIPTION
**Pull request by ianwremmel**
Closes #615
Closes #602
Closes #706

### CODE REVIEW

File: packages/node_modules/@ciscospark/media-engine-webrtc/src/engine.js [hunk 1]
```diff
@@ -0,0 +1,605 @@
...
+const DirectionContainer = makeTemplateContainer(WeakMap, Map);
+const targetMediaDirection = new DirectionContainer();
+
+const capitalize = {
+  audio: `Audio`,
+  video: `Video`
+};
+
+/**
+ * Wrapper around targetMediaDirection.get which returne `inactive` instead of
```
Comment (adamweeks) [inline packages/node_modules/@ciscospark/media-engine-webrtc/src/engine.js]: spelling: `returne`

File: packages/node_modules/@ciscospark/media-engine-webrtc/src/webrtc-adapter-adapter.js [hunk 2]
```diff
@@ -0,0 +1,51 @@
+/**
+ * @file webrtc-adapter-adapter
+ * Yep, you read that filename right. There seems to be a bug in adapter.js that
+ * doesn't quite clean up everythings it supposed to. This is an adapter for the
+ * adapter that makes sure it removes the correct streams from the
+ * PeerConnection's caches
+ */
+
+if (!process.env.NO_WEBRTC_ADAPTER) {
```
Comment (adamweeks) [inline packages/node_modules/@ciscospark/media-engine-webrtc/src/webrtc-adapter-adapter.js]: 👍

File: packages/node_modules/@ciscospark/media-engine-webrtc/src/webrtc-helpers.js [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+  }
+
+  let send = false;
+  const senders = pc
+    .getSenders()
+    .filter((s) => s.track.kind === kind);
+
+  send = senders.length > 0 && senders.reduce((acc, s) => acc && s.track.enabled, true);
+
+  // let recv = false;
```
Comment (adamweeks) [inline packages/node_modules/@ciscospark/media-engine-webrtc/src/webrtc-helpers.js]: Supposed to be commented out?

Comment (ianwremmel) [reply packages/node_modules/@ciscospark/media-engine-webrtc/src/webrtc-helpers.js]: sort of. it's there in the hopes that some day chrome will do receivers correctly. we shouldn't be using get*Streams anywhere, but Chrome and/or the webrtc-adapter still hasn't quite wired the receivers array up to to be accurate.

File: packages/node_modules/@ciscospark/media-engine-webrtc/test/unit/spec/engine.js [hunk 4]
```diff
@@ -0,0 +1,776 @@
...
+                before(() => {
+                  engine = new WebRTCMediaEngine();
+                  engine.on(`change:receivingVideo`, (value, previous) => console.log(`XXXXX receivingVideo`, engine.receivingVideo, value, previous));
+                });
+
+                after(() => {
+                  engine.stop();
+                });
+
+                step(`initiates a session with audio=${audioStart} and video=${videoStart}`, () => {
```
Comment (adamweeks) [inline packages/node_modules/@ciscospark/media-engine-webrtc/test/unit/spec/engine.js]: `step`! Nice!

File: packages/node_modules/@ciscospark/media-engine-webrtc/src/webrtc-helpers.js (hunk 3 above)
Comment (adamweeks) [reply packages/node_modules/@ciscospark/media-engine-webrtc/src/webrtc-helpers.js]: I gotcha, could you put a note in there about it?

File: packages/node_modules/@ciscospark/media-engine-webrtc/test/unit/spec/engine.js [hunk 5]
```diff
@@ -0,0 +1,776 @@
...
+      assert.lengthOf(senders, 1, `local ${kind} is ${direction}`);
+      assert.isFalse(tracks[0].enabled, `the local ${kind} track is not enabled`);
+      assert.isFalse(senders[0].track.enabled, `local ${kind}is ${direction}`);
+    }
+  }
+}
+
+function assertRemoteKind(kind, direction, engine) {
+  const tracks = engine.remoteMediaStream.getTracks().filter((t) => t.kind === kind);
+  // const pcTracks = engine.pc.getRemoteStreams[0] && engine.pc.getRemoteStreams[0].getTracks().filter((t) => t.kind === kind);
```
Comment (adamweeks) [inline packages/node_modules/@ciscospark/media-engine-webrtc/test/unit/spec/engine.js]: `pcTracks` tests supposed to be commented?

**Comment by ianwremmel** [comment]
Well, turns out it still needs some work; we can't create the local and remote media streams until we have tracks to put in them. (if you try to attach them to a video element, you get an error

**Comment by ianwremmel** [comment]
Something in here is really upsetting Firefox and I can't tell what. It seems like there are specific scenarios where leaving a call may cause Firefox to hang, but I don't think it's a real-world use case, simply a result of the way we run tests. Plans for now: 
- [ ] lower the browser disconnect timeout and increase the browser disconnect tolerance
- [ ] make sure all phone tests hangup their own calls (relying an hangup via spies didn't work for some reason; something with referencing the spy's `returnValues` caused firefox to hang)
- [ ] make all tests pass in Firefox (once tests hangup correctly, chrome should start passing again)
- [ ] fix plugin phone unit tests

File: packages/node_modules/@ciscospark/media-engine-webrtc/test/unit/spec/engine.js (hunk 4 above)
Comment (adamweeks) [reply packages/node_modules/@ciscospark/media-engine-webrtc/test/unit/spec/engine.js]: @moriahmaney said that step and retries don't play nice, may need to confirm this one

Comment (ianwremmel) [reply packages/node_modules/@ciscospark/media-engine-webrtc/test/unit/spec/engine.js]: yea, i could see that, but with the ways these are staged, retries wouldn't help anyway.

File: packages/node_modules/@ciscospark/media-engine-webrtc/test/unit/spec/engine.js (hunk 5 above)
Comment (ianwremmel) [reply packages/node_modules/@ciscospark/media-engine-webrtc/test/unit/spec/engine.js]: yea, i think maybe there's a comment in assertLocalKind that didn't make it here...

File: packages/node_modules/@ciscospark/plugin-phone/test/integration/lib/event-expectations.js [hunk 6]
```diff
@@ -0,0 +1,61 @@
+import {expectEvent} from '@ciscospark/test-helper-mocha';
+
+export function expectCallIncomingEvent(ctx, msg) {
```
Comment (adamweeks) [inline packages/node_modules/@ciscospark/plugin-phone/test/integration/lib/event-expectations.js]: This is a neat helper library!

File: packages/node_modules/@ciscospark/media-engine-webrtc/src/engine.js [hunk 7]
```diff
@@ -0,0 +1,691 @@
...
+ */
+export default class WebRTCMediaEngine {
+  /**
+   * Wrapper around navigator.mediaDevices.getUserMedia
+   *
+   * @param {MediaStreamContraints} constraints
+   * @returns {Promise<MediaStream>}
+   */
+  static getUserMedia(constraints) {
+    const finalConstraints = defaults({}, constraints, {fake: process.env.NODE_ENV === `test`});
```
Comment (adamweeks) [inline packages/node_modules/@ciscospark/media-engine-webrtc/src/engine.js]: How does this work with Chrome's fake media flags? Do we need to set both the flags and the NODE_ENV?

Comment (ianwremmel) [reply packages/node_modules/@ciscospark/media-engine-webrtc/src/engine.js]: this only works in firefox; in chrome, you need to launch with the appropriate switches.

**Pull request merged**